### PR TITLE
Walks

### PR DESCRIPTION
**Pull request by vassilit**
This is #17 and #18 combined into one branch.

### CODE REVIEW

**Comment by emersion** [comment]
I don't want `Separator` and `ReaddirChunk` to be part of the public API.

**Comment by vassilit** [comment]
Bonjour emersion,

In the comment for separator,

```
// This should only be changed on operating systems where the colon isn't
// allowed in filenames.

```
Should we understand «change» as «fork go-maildir and use maintain you own branch» ?

It is common for packages to export global variables (such as timeouts, etc) so that most users will use the defaults, but some users might want to tune some aspects of the package. It's not ideal (thread unsafe if misused), but convenient nevertheless. 

For example, if I'm using FAT32, I want to be able to change the separator. If I have a lot of RAM, I'd better have ReaddirChunk set at 4096. Otherwise, the default (100) is good for most users.

What do you suggest so that maildirpp can use the same readdirChunk value as its parent package, maildir?

**Comment by emersion** [comment]
I don't think that comment makes sense. Using something other than a colon would break interoperability with other software, and the spec doesn't mention anything about it.

Global variables are not a good way to configure packages. Per-struct options are a better design.

I would rather not make internal details such as separator and readdir chunk configurable unless there is a solid real use-case for it. "Someone might need it" is not good enough for me, as often times we end up designing a broken configuration option for an imaginary use-case which doesn't end up working in practice.

I don't think it's very important to have a single readdir chunk variable, I think copying it over is fine.

**Comment by emersion** [comment]
Ref this comment which has not been entirely addressed: https://github.com/emersion/go-maildir/pull/18#issuecomment-1571520550

**Comment by vassilit** [comment]
> Nit: the Keys function could be implemented based on the new Walk function.

IMHO, no need to process flags (as Dir.Walk does) when only keys are needed.

**Comment by vassilit** [comment]
> Nit: the Keys function could be implemented based on the new Walk function.

IMHO, no need to process flags (as Dir.Walk does) when only keys are needed.

File: maildir.go [hunk 1]
```diff
@@ -502,6 +500,44 @@ func (d Dir) Clean() error {
...
+
+	for {
+		names, err := f.Readdirnames(readdirChunk)
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		for _, filename := range names {
```
Comment (emersion) [inline maildir.go:523]: We should skip filenames starting with ".", I think.

File: maildir.go [hunk 2]
```diff
@@ -502,6 +500,44 @@ func (d Dir) Clean() error {
...
+			key, err := parseKey(filename)
+			if err != nil {
+				return err
+			}
+			flags, err := d.Flags(key)
```
Comment (emersion) [inline maildir.go:528]: We should be able to get the flags from the filename directly, without needing to go back to guessing the filename from the key in `Dir.Flags`. I'd suggest extracting some of the `Dir.Flags` logic into a `parseFlags(filename string) ([]Flag, error)` function.

File: maildir.go [hunk 3]
```diff
@@ -502,6 +500,44 @@ func (d Dir) Clean() error {
 	return nil
 }
 
+// Walk calls fn for every messages in the mailbox's subdir directory.
+//
+// subdir should be "cur", "new" or "tmp".
```
Comment (emersion) [inline maildir.go:505]: I'm not comfortable exposing the subdir directly to the API user. I think only iterating over "cur" is desirable?

**Comment by emersion** [comment]
> IMHO, no need to process flags (as Dir.Walk does) when only keys are needed.

If we compute directly the flags from the filename, then it's a very cheap operation, and I don't think it's worth avoiding its cost.

**Comment by emersion** [comment]
Superseded by 69317e079e05f380c56d419e6600c02cf97cec8f

**Comment by vassilit** [comment]
> > // This should only be changed on operating systems where the colon isn't
> > // allowed in filenames.
> I don't think that comment makes sense. Using something other than a colon would break interoperability with other software, and the spec doesn't mention anything about it.

There are some programs that support VFAT-based Maildirs, which use '!' as the separators instead of ':' and don't use '.' at the beginning of a directory name. I agree that it is not the role of go-maildir to support this special format.

Example: [Tinymail](https://github.com/maemo-leste/tinymail/blob/b008eb1ab858cf54186bf703eab385fe9a8ed602/ChangeLog#L3412)

**Pull request closed without merging**